File: si-gui-desktop/src/main/java/at/reisisoft/sigui/ui/MainUI.java
package at.reisisoft.sigui.ui;

import at.reisisoft.sigui.settings.SettingsKt;
import at.reisisoft.sigui.settings.SiGuiSetting;
import at.reisisoft.ui.JavaFXKt;
import at.reisisoft.ui.JavaFxUtils;
import javafx.application.Application;
import javafx.collections.ObservableList;
import javafx.fxml.FXMLLoader;
import javafx.scene.Parent;
import javafx.scene.Scene;
import javafx.scene.image.Image;
import javafx.stage.Stage;
import kotlin.collections.CollectionsKt;

import java.io.IOException;
import java.io.InputStream;
import java.io.UncheckedIOException;
import java.nio.charset.StandardCharsets;
import java.util.Locale;
import java.util.Objects;
import java.util.ResourceBundle;

public class MainUI extends Application {

    private MainUIController controller;

    @Override
    public void start(Stage primaryStage) throws Exception {
        Thread.currentThread().setUncaughtExceptionHandler(JavaFXKt.getUNCAUGHT_EXCEPTION_HANDLER());

        ObservableList<Image> icons = primaryStage.getIcons();
        CollectionsKt.listOf("48", "72", "96", "144", "192").forEach(it -> {
            try (InputStream icon = MainUI.class.getClassLoader().getResource("icons/" + it + ".png").openStream()) {
                icons.add(new Image(icon));
            } catch (IOException e) {
                throw new UncheckedIOException(e);
            }
        });

        SiGuiSetting settings = SettingsKt.loadSettings();
        Locale.setDefault(settings.getUiLanguage());
        FXMLLoader loader = JavaFxUtils.loadFXML("mainUI.fxml");
        ResourceBundle languageSupport = JavaFXKt.loadEncodedRessource("uistrings.sigui-desktop", StandardCharsets.UTF_8);
        loader.setResources(languageSupport);
        primaryStage.setTitle(languageSupport.getString(ResourceKey.APPNAME.toString()));
        Parent mainUi = loader.load();
        controller = loader.getController();
        controller.internalInitialize(settings, primaryStage);
        primaryStage.setScene(new Scene(mainUi));
        primaryStage.show();
    }

    public static void main(String[] args) {
        Thread.currentThread().setUncaughtExceptionHandler(JavaFXKt.getUNCAUGHT_EXCEPTION_HANDLER());
        launch(args);
    }

    @Override
    public void stop() {
        Objects.requireNonNull(controller).close();
        System.exit(0);
    }
}

File: si-gui-desktop/src/main/kotlin/at/reisisoft/sigui/ui/MainUI.kt
package at.reisisoft.sigui.ui

import at.reisisoft.sigui.settings.loadSettings
import at.reisisoft.ui.*
import javafx.application.Application
import javafx.scene.Scene
import javafx.scene.image.Image
import javafx.stage.Stage
import java.nio.charset.StandardCharsets
import java.util.*

class MainUI : Application() {

    private lateinit var controller: MainUIController


    override fun start(primaryStage: Stage) {
        Thread.currentThread().uncaughtExceptionHandler = UNCAUGHT_EXCEPTION_HANDLER

        val icons = primaryStage.icons
        listOf("16", "32", "48", "64", "72", "96", "144", "192", "512").forEach { it ->
            MainUI::class.java.classLoader.getResource("icons/$it.png")?.openStream()?.use {
                icons += Image(it)
            } ?: throw IllegalStateException("Missing icon of size: ${it}x$it")

        }

        val settings = loadSettings()
        Locale.setDefault(settings.uiLanguage)
        JavaFxUtils.loadFXML("mainUI.fxml").apply {
            loadEncodedRessource("uistrings.sigui-desktop", StandardCharsets.UTF_8).let {
                resources = it!!
                primaryStage.title = it.getString(ResourceKey.APPNAME)
            }
            loadAsParent().let { mainUi ->
                controller = getController()
                controller.setData(settings, primaryStage)
                primaryStage.scene = Scene(mainUi)
                primaryStage.show()
            }
        }
    }

    companion object {
        @JvmStatic
        fun main(args: Array<String>) {
            Thread.currentThread().uncaughtExceptionHandler = UNCAUGHT_EXCEPTION_HANDLER
            launch(MainUI::class.java, *args)
        }
    }

    override fun stop() {
        Objects.requireNonNull(controller).close()
        System.exit(0)
    }
}